File: 0300_Longest Increasing Subsequence/LongestIncreasingSubsequence_BinarySearch.cs
public class Solution {
    public int LengthOfLIS(int[] nums) {
        if(nums.Length == 1) return 1;
        var dp = new List<int>();
        for(int i=0;i<nums.Length;i++)
        {
            var pos = BinarySearch(dp,nums[i]);
            if(pos == dp.Count)
                dp.Add(nums[i]);
            else
                dp[pos] = nums[i];
        }
        
        return dp.Count();
    }
    
    private int BinarySearch(List<int> list, int target)
    {
        var l = 0;
        var r = list.Count;
        while(l < r)
        {
            var mid = (r-l)/2 + l;
            if(list[mid] >= target)
                r = mid;
            else
                l = mid+1;
        }
        
        return l;
    }
}

File: 0300_Longest Increasing Subsequence/LongestIncreasingSubsequence_BinarySearch.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> list = new ArrayList<>();
        
        for(int i=0;i<nums.length;i++) {
            int pos = lowerBound(list, nums[i]);
            if(pos == list.size()) {
                list.add(nums[i]);
            }else {
                list.set(pos, nums[i]);
            }
        }
        
        return list.size();
    }
    
    private int lowerBound(List<Integer> list, int target) {
        int l = 0;
        int r = list.size();
        
        while(l < r) {
            int mid = (r-l)/2 + l;
            if(list.get(mid) < target) {
                l = mid + 1;
            }else {
                r = mid;
            }
        }
        
        return l;
    }
}